Add support for URL parameters

lohmander opened the pull request:
As briefly discussed in #10. 

I didn't actually implement regex support, but rather plain `:param` style parameters. I'm passing the values as the third argument to the handler function. Maybe a better option would be to pass a request object, including both the body, params, headers, etc.

Commit:
Add support for URL parameters

Commit:
Make all lodash imports in a single import (since we use chain from the baselib anyways)

Commit:
Update README to include URL params

Commit:
Fix lint errors

muuki88 commented on src/routing.js:
```diff
@@ -0,0 +1,76 @@
+import {chain, merge, toPairs, trim} from 'lodash';
+
+const isParam = segment => segment.substr(0, 1) === ':';
```
I think `startsWith` is clearer than `substr(0,1`.

muuki88 commented on src/routing.js:
```diff
@@ -0,0 +1,76 @@
+const isParam = segment => segment.substr(0, 1) === ':';
+const toSegments = pathname => trim(pathname, '/').split('/');
+
+const buildRoute = (segments, methods) => {
```
Can you add some documentation to this method, e.g.

- Short explanation what this method does
- What types should the input parameters have
- What's the return type

muuki88 commented on src/routing.js:
```diff
@@ -0,0 +1,76 @@
+
+const buildRoute = (segments, methods) => {
+  const segment = segments.shift();
+  const edge = segments.length === 0;
```
I would rather use `isEdge` to indicate that this is a bool value.
`isLastSegment` is the better name here ( as you used below )

muuki88 commented on src/routing.js:
```diff
@@ -0,0 +1,76 @@
+const toSegments = pathname => trim(pathname, '/').split('/');
+
+const buildRoute = (segments, methods) => {
+  const segment = segments.shift();
```
`shift` mutates the `segments` array, right? I think we should keep these mutation / side effects out of a function.
Try using the [array spread operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment) instead, e.g.

```js
const [nextSegment, ...remainingSegments] = segments;
const isLastSegment = !(remainingSegments && remainingSegments.length > 0);
```

muuki88 commented on src/routing.js:
```diff
@@ -0,0 +1,76 @@
+
+const lookup = (segments, routes) => {
+  const segment = segments.shift();
+  const isLastSegment = segments.length === 0;
```
The same as above. Use array spreading instead of shift to avoid mutation.

muuki88 commented on src/routing.js:
```diff
@@ -0,0 +1,76 @@
+const buildParams = (param, value) =>
+  param ? { params: { [param.substr(1)]: value } } : {};
+
+const lookup = (segments, routes) => {
```
Please add some documentation for this method. What are `segments` and `routes` types and what are the expected return types.

muuki88 reviewed (changes requested):
Thanks a lot for your good work on this @lohmander :heart: :smiley: 

I made a few comments, that should be trivial to fix or be discussed about :)

Very well done pull request with description, tests and updated documentation :+1:

lohmander commented:
All sound points, I'll look into fixing everything next week. 👍

On Jul 1, 2017 11:30 AM, "Nepomuk Seiler" <notifications@github.com> wrote:

> *@muuki88* requested changes on this pull request.
>
> Thanks a lot for your good work on this @lohmander
> <https://github.com/lohmander> ❤️ 😃
>
> I made a few comments, that should be trivial to fix or be discussed about
> :)
>
> Very well done pull request with description, tests and updated
> documentation 👍
> ------------------------------
>
> In src/routing.js
> <https://github.com/gutefrage/redux-effects-fetch-fixture/pull/11#discussion_r125158912>
> :
>
> > @@ -0,0 +1,76 @@
> +import {chain, merge, toPairs, trim} from 'lodash';
> +
> +const isParam = segment => segment.substr(0, 1) === ':';
>
> I think startsWith is clearer than substr(0,1.
> ------------------------------
>
> In src/routing.js
> <https://github.com/gutefrage/redux-effects-fetch-fixture/pull/11#discussion_r125158921>
> :
>
> > @@ -0,0 +1,76 @@
> +import {chain, merge, toPairs, trim} from 'lodash';
> +
> +const isParam = segment => segment.substr(0, 1) === ':';
> +const toSegments = pathname => trim(pathname, '/').split('/');
> +
> +const buildRoute = (segments, methods) => {
>
> Can you add some documentation to this method, e.g.
>
>    - Short explanation what this method does
>    - What types should the input parameters have
>    - What's the return type
>
> ------------------------------
>
> In src/routing.js
> <https://github.com/gutefrage/redux-effects-fetch-fixture/pull/11#discussion_r125158963>
> :
>
> > @@ -0,0 +1,76 @@
> +import {chain, merge, toPairs, trim} from 'lodash';
> +
> +const isParam = segment => segment.substr(0, 1) === ':';
> +const toSegments = pathname => trim(pathname, '/').split('/');
> +
> +const buildRoute = (segments, methods) => {
> +  const segment = segments.shift();
> +  const edge = segments.length === 0;
>
> I would rather use isEdge to indicate that this is a bool value.
> isLastSegment is the better name here ( as you used below )
> ------------------------------
>
> In src/routing.js
> <https://github.com/gutefrage/redux-effects-fetch-fixture/pull/11#discussion_r125159102>
> :
>
> > @@ -0,0 +1,76 @@
> +import {chain, merge, toPairs, trim} from 'lodash';
> +
> +const isParam = segment => segment.substr(0, 1) === ':';
> +const toSegments = pathname => trim(pathname, '/').split('/');
> +
> +const buildRoute = (segments, methods) => {
> +  const segment = segments.shift();
>
> shift mutates the segments array, right? I think we should keep these
> mutation / side effects out of a function.
> Try using the array spread operator
> <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment>
> instead, e.g.
>
> const [nextSegment, ...remainingSegments] = segments;const isLastSegment = !(remainingSegments && remainingSegments.length > 0);
>
> ------------------------------
>
> In src/routing.js
> <https://github.com/gutefrage/redux-effects-fetch-fixture/pull/11#discussion_r125159118>
> :
>
> > +  return {
> +    [segment]: merge(
> +      { $param: isParam(segment) },
> +      edge
> +        ? { $edge: edge, $methods: methods }
> +        : buildRoute(segments, methods)
> +    )
> +  };
> +};
> +
> +const buildParams = (param, value) =>
> +  param ? { params: { [param.substr(1)]: value } } : {};
> +
> +const lookup = (segments, routes) => {
> +  const segment = segments.shift();
> +  const isLastSegment = segments.length === 0;
>
> The same as above. Use array spreading instead of shift to avoid mutation.
> ------------------------------
>
> In src/routing.js
> <https://github.com/gutefrage/redux-effects-fetch-fixture/pull/11#discussion_r125159158>
> :
>
> > +  const edge = segments.length === 0;
> +
> +  return {
> +    [segment]: merge(
> +      { $param: isParam(segment) },
> +      edge
> +        ? { $edge: edge, $methods: methods }
> +        : buildRoute(segments, methods)
> +    )
> +  };
> +};
> +
> +const buildParams = (param, value) =>
> +  param ? { params: { [param.substr(1)]: value } } : {};
> +
> +const lookup = (segments, routes) => {
>
> Please add some documentation for this method. What are segments and
> routes types and what are the expected return types.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gutefrage/redux-effects-fetch-fixture/pull/11#pullrequestreview-47527073>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACGGOuUlyZwh7K1AyrfHvq4fKyVELw8Sks5sJhGfgaJpZM4OKfaC>
> .
>

Commit:
Add some code docs and avoid mutations

lohmander commented:
@muuki88 I think all of the issues should be fixed now.

muuki88 reviewed (approved):
Awesome! Thanks a lot. I'll release this ASAP
